### PR TITLE
[INTEL MKL] Enabling MKL support for FusedBatchNormV3

### PR DESCRIPTION
**Pull request by jojivk73**
Adding Intel MKL support for FusedBatchNormV3

### CODE REVIEW

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc [hunk 1]
```diff
@@ -818,10 +829,29 @@ class MklFusedBatchNormOp : public OpKernel {
...
+    if (reserved_space) {
+      DCHECK(reserved_space_tensor != nullptr);
+
+      MklDnnShape mkl_shape_reserved_space;
+      mkl_shape_reserved_space.SetMklTensor(false);
+      AllocateOutputSetMklShape(context, kReservedSpaceIndex,
+                                reserved_space_tensor, tf_shape_scale,
+                                mkl_shape_reserved_space);
+      DCHECK((*reserved_space_tensor) != nullptr);
+      // set NAN variance value in case of empty input tensor
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: This is not variance value, please correct the description. 
Nit: Please also capitalize the first letter of the sentence.

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc [hunk 2]
```diff
@@ -843,13 +873,17 @@ class MklFusedBatchNormGradOp : public OpKernel {
       const size_t kScaleIndex = 2;     // index of scale tensor
       const size_t kMeanIndex = 3;      // index of saved_mean tensor
       const size_t kVarianceIndex = 4;  // index of saved_variance tensor
+      const size_t kReSpaceIndex = 5;   // index of reserved space 3 tensor
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Please type the name in full for readability.
```suggestion
      const size_t kReservedSpaceIndex = 5;   // index of reserved space 3 tensor
```

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc [hunk 3]
```diff
@@ -843,13 +873,17 @@ class MklFusedBatchNormGradOp : public OpKernel {
...
 
       const Tensor& diff_dst_tensor = MklGetInput(context, kDiffDstIndex);
       const Tensor& src_tensor = MklGetInput(context, kSrcIndex);
       const Tensor& scale_tensor = MklGetInput(context, kScaleIndex);
       const Tensor& saved_mean_tensor = MklGetInput(context, kMeanIndex);
       const Tensor& saved_variance_tensor =
           MklGetInput(context, kVarianceIndex);
+      const Tensor temp;
+      const Tensor& reserved_space_tensor =
+          (reserved_space) ? MklGetInput(context, kReSpaceIndex) : temp;
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Can we replace `temp` with `Tensor()`?
```suggestion
          (reserved_space) ? MklGetInput(context, kReSpaceIndex) : Tensor();
```

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc [hunk 4]
```diff
@@ -1014,9 +1048,16 @@ class MklFusedBatchNormGradOp : public OpKernel {
       U* weights_data = weights_data_tf;
       T* diff_src_data = static_cast<T*>(diff_src_tensor->flat<T>().data());
       U* diff_weights_data = static_cast<U*>(diff_weights.GetAllocatedBuffer());
+
+      U* res_space_data = nullptr;
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: You can make this a constant using the ternary clause.
```C++
const U* res_space_data = (reserved_space)? casting_of_actual_pointer : nullptr;
```
In general, if any variable is not modified in its lifetime, it should be a constant.

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 5]
```diff
@@ -1140,6 +1140,72 @@ MKL-DNN implementation of quantized depthwise Conv2D with Bias, Relu and Requant
...
+    .Output("reserve_space_1: U")
+    .Output("reserve_space_2: U")
+    .Output("reserve_space_3: U")
+    .Output("mkl_y: uint8")
+    .Output("mkl_batch_mean: uint8")
+    .Output("mkl_batch_variance: uint8")
+    .Output("mkl_reserve_space_1: uint8")
+    .Output("mkl_reserve_space_2: uint8")
+    .Output("mkl_reserve_space_3: uint8")
+    .Attr("T: {bfloat16, float}")
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc]: The normal TF version also supports `T = half`. Do you need to check for `half` during graph rewrite? Same for the grad op.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR!

File: tensorflow/core/graph/mkl_layout_pass.cc [hunk 6]
```diff
@@ -2961,6 +2979,14 @@ void MklLayoutRewritePass::CopyAttrsFusedBatchNormV2(const Node* orig_node,
   nb->Attr("U", U);
 }
 
+// Calls corresp api of FBNV2, for V3 on CPU, as there are no additional
+// attributes.
+void MklLayoutRewritePass::CopyAttrsFusedBatchNormV3(const Node* orig_node,
```
[the remark this answers is not in the document]
Comment (jojivk73) [reply tensorflow/core/graph/mkl_layout_pass.cc]: Deleted

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc (hunk 1 above)
Comment (jojivk73) [reply tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Done

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc (hunk 3 above)
Comment (jojivk73) [reply tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Done

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 5 above)
Comment (jojivk73) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc (hunk 4 above)
Comment (jojivk73) [reply tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Done

**Comment by jojivk73** [comment]
Made the changes to the PR as suggested!
Thanks

**Review by penpornk — APPROVED** [review]
Thank you for your quick fixes!

**Pull request merged**